### PR TITLE
SWATCH-2410: Add v2 of the subscriptions/products API

### PR DESCRIPTION
**Pull request by Sgitario**
Jira issue: SWATCH-2410

## Description
The old v1 API returns:

```json
{
  "data": [
    {
      "sku": "ESA0012",
      "product_name": "Red Hat OpenShift Container Platform, Premium (One Year, Enterprise Program)",
      "service_level": "Premium",
      "usage": "",
      "subscriptions": [
        {
          "id": "14349691",
          "number": "14349848"
        }
      ],
      "billing_provider": "",
      "next_event_date": "2025-02-19T05:00:00Z",
      "next_event_type": "Subscription End",
      "quantity": 1,
      "capacity": 0,
      "hypervisor_capacity": 0,
      "total_capacity": 0,
      "has_infinite_quantity": true
    }
  ],
  "meta": {
    "count": 1,
    "product": "OpenShift Container Platform",
    "subscription_type": "Annual"
  }
}
```

Where the new v2 API will always populate the metrics that are configured in the product:

```json
{
  "data": [
    {
      "sku": "ESA0012",
      "product_name": "Red Hat OpenShift Container Platform, Premium (One Year, Enterprise Program)",
      "service_level": "Premium",
      "usage": "",
      "subscriptions": [
        {
          "id": "14349691",
          "number": "14349848"
        }
      ],
      "billing_provider": "",
      "next_event_date": "2025-02-19T05:00:00Z",
      "next_event_type": "Subscription End",
      "quantity": 1,
      "measurements": [
        0.0,
        0.0
      ],
      "has_infinite_quantity": true
    }
  ],
  "meta": {
    "count": 1,
    "product": "OpenShift Container Platform",
    "measurements": [
      "Sockets",
      "Cores"
    ],
    "subscription_type": "Annual"
  }
}
```

Changes in the new API:
- added the same format as in the Instances API: "item.measurements", "meta.measurements", "item.category".
- "capacity" and "hypervisor_capacity" is now "measurements" which is a list of numbers. The metricId is defined in the meta.measurements object.
- Since one offering can be either hypervisor or not, we no longer provide both "capacity" and "hypervisor_capacity", but only the "measurements" and the "category" object which is of type ReportCategory. 
- Removed "total_capacity" since this was a sum of the "capacity" and "hypervisor_capacity" values which as stated above, it was removed.

### Setup
1.- deploy the app into an EE
2.- create a contract via API. Instructions using the iqe tooling:

```
import random
product_id = "rosa"

// create contract:
contract_data = {
    "instance_hours": random.randint(0, 10),
    "cores": random.randint(0, 7),
    "billing_provider": "aws",
    "billing_account_id": "AAA",
}
contract = app.rhsm_subscriptions.create_contract(
    product_id=product_id,
    **contract_data,
)
```

3.- confirm that the above API creates two subscriptions
4.- create measurements for both subscriptions (there is one that does not have any). For example:

```
insert into subscription_measurements(subscription_id, start_date, metric_id, value, measurement_type)
values ('488052','2024-05-16 10:01:08.000 +0200', 'Cores', '50', 'PHYSICAL');
```

5.- confirm the capacity that is returned from the v2 subscription API.

Assuming that the cores values for the subscriptions are 10 and 50 and org ID is 3340851:

```
echo -n '{"identity":{"account_number":"","type":"User","user":{"is_org_admin":true},"internal":{"org_id":"3340851"}}}' | base64 -w 0
-- eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6IiIsInR5cGUiOiJVc2VyIiwidXNlciI6eyJpc19vcmdfYWRtaW4iOnRydWV9LCJpbnRlcm5hbCI6eyJvcmdfaWQiOiIzMzQwODUxIn19fQ==
```

```
curl -X 'PUT' \
  'http://localhost:8000/api/rhsm-subscriptions/v1/opt-in' \
  -H 'accept: application/vnd.api+json' \
  -H 'x-rh-identity: eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6IiIsInR5cGUiOiJVc2VyIiwidXNlciI6eyJpc19vcmdfYWRtaW4iOnRydWV9LCJpbnRlcm5hbCI6eyJvcmdfaWQiOiIzMzQwODUxIn19fQ=='
```

```
curl -X GET \
  'http://localhost:8000/api/rhsm-subscriptions/v2/subscriptions/products/rosa' \
  -H 'accept: application/vnd.api+json' \
  -H 'x-rh-identity: eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6IiIsInR5cGUiOiJVc2VyIiwidXNlciI6eyJpc19vcmdfYWRtaW4iOnRydWV9LCJpbnRlcm5hbCI6eyJvcmdfaWQiOiIzMzQwODUxIn19fQ==' | jq
```

You should see the new format as the output.

### CODE REVIEW

**Comment by cdcabrera** [comment]
@Sgitario response looks like something the UI will have no issues consuming. We do have a few questions we posted in the Jira side of things

**Comment by Sgitario** [comment]
PR updated by using the same view as used in the Export of the subscriptions objects. So, with these changes, both API and export work the same.
In addition to this, I've also refactored the view to increase the overall performance.

**Comment by Sgitario** [comment]
the view's changed will be done in a separated PR: https://github.com/RedHatInsights/rhsm-subscriptions/pull/3432
keeping this pull request as it is and will rebase after it's merged.

**Comment by Sgitario** [comment]
I will resolve the conflicts once https://github.com/RedHatInsights/rhsm-subscriptions/pull/3432 is merged

**Comment by Sgitario** [comment]
Putting the PR on draft since the changes strongly depend on https://github.com/RedHatInsights/rhsm-subscriptions/pull/3452.

File: swatch-core/src/main/java/org/candlepin/subscriptions/db/model/BillingProvider.java
Comment (kartikshahc) [inline swatch-core/src/main/java/org/candlepin/subscriptions/db/model/BillingProvider.java:31]: Why would we want it `__EMPTY__` instead of EMPTY?

File: src/test/java/org/candlepin/subscriptions/subscription/SubscriptionTableControllerOnDemandTest.java
Comment (kartikshahc) [inline src/test/java/org/candlepin/subscriptions/subscription/SubscriptionTableControllerOnDemandTest.java:40]: Why do we have v1 here?

**Review by kartikshahc — CHANGES_REQUESTED** [review]
Left a few questions here. Since I will be OOO, I think this has to be assigned to a different dev for a review so you are not blocked.

File: src/main/java/org/candlepin/subscriptions/resource/api/v1/SubscriptionResource.java
Comment (kartikshahc) [inline src/main/java/org/candlepin/subscriptions/resource/api/v1/SubscriptionResource.java:34]: Why are we adding Uom to v1? I thought we are removing it overall.

File: swatch-core/src/main/java/org/candlepin/subscriptions/db/model/BillingProvider.java
Comment (Sgitario) [reply swatch-core/src/main/java/org/candlepin/subscriptions/db/model/BillingProvider.java:31]: This type is now generated by json schema generator instead of openapi, and the enum name changed.

**Comment by Sgitario** [comment]
@kartikshahc I'm not removing the uom from the v1 (this was not planned for this PR).

File: src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionTableController.java
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionTableController.java:251]: Soft suggestion: what do you think of using mapstruct here? Given the field names are mostly the same, it might reduce some boilerplate.

File: src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionTableController.java
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionTableController.java:107]: Rather than encoding into a string like this, I'd rather see a small data class having sku and billing provider as fields.

**Review by kahowell — CHANGES_REQUESTED** [review]

File: api/rhsm-subscriptions-api-v1-spec.yaml
Comment (Sgitario) [reply api/rhsm-subscriptions-api-v1-spec.yaml:143]: I initially did this, but found that we had to duplicate a lot of logic just because these types are now v1 and v2. 
Anyway, it makes sense to me. I will re-evaluate this and we can check if it's worthy or not.

File: src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionResource.java
Comment (Sgitario) [reply src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionResource.java:41]: I think this is a spring limitation, that's why prefered overwritting the bean name using the `@Component` annotation. 
Yet, I don't have a strong oppinion here, so if you see more intuitive the class renaming, will do it.

File: api/rhsm-subscriptions-api-v1-spec.yaml
Comment (Sgitario) [reply api/rhsm-subscriptions-api-v1-spec.yaml:143]: Update: after taking over the review, I now recalled why I had to make the model common: we have some entities that directly use the model API. So, we can't map these entities to multiple versioned model APIs. 
Example: https://github.com/RedHatInsights/rhsm-subscriptions/blob/main/swatch-core/src/main/java/org/candlepin/subscriptions/db/model/Usage.java

File: api/rhsm-subscriptions-api-v2-spec.yaml
Comment (Sgitario) [reply api/rhsm-subscriptions-api-v2-spec.yaml]: updated

File: src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionResource.java
Comment (Sgitario) [reply src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionResource.java]: updated

File: src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionTableController.java
Comment (Sgitario) [reply src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionTableController.java]: updated

File: src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionTableController.java
Comment (lindseyburnett) [reply src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionTableController.java:249]: +1 to the MapStruct suggestion

File: api/rhsm-subscriptions-api-v1-spec.yaml
Comment (kahowell) [reply api/rhsm-subscriptions-api-v1-spec.yaml]: Resolved in latest commit

**Review by lindseyburnett — APPROVED** [review]

**Comment by kahowell** [comment]
@mmusil FYI, this is in its final form from a dev perspective, eager to get your feedback.

**Review by lindseyburnett — APPROVED** [review]

**Comment by mmusil** [comment]
I deployed the latest images to EE, and re-generated API. So far, it's looking good.
RHEL:
```
In [22]: app.rhsm_subscriptions.get_skus_report(product_id)
2024-07-19 07:21:24.515 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=None, params=[('sort', 'next_event_date'), ('limit', 100), ('offset', 0)]
Out[22]: 
[{'billing_provider': '',
  'capacity': 0,
  'has_infinite_quantity': True,
  'hypervisor_capacity': 0,
  'next_event_date': datetime.datetime(2024, 8, 3, 3, 59, 59, tzinfo=tzlocal()),
  'next_event_type': 'Subscription End',
  'product_name': 'Red Hat Enterprise Linux, Premium (One Year, Enterprise '
                  'Program)',
  'quantity': 1,
  'service_level': 'Premium',
  'sku': 'ESA0001',
  'subscriptions': [{'id': '14462076', 'number': '14462233'}],
  'total_capacity': 0,
  'usage': 'Production'}]


In [23]: app.rhsm_subscriptions.rest_client.subscriptions_api.get_sku_capacity_report(product_id)
2024-07-19 07:21:36.438 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=None, params=[]
Out[23]: 
{'data': [{'billing_provider': '',
           'capacity': 0,
           'has_infinite_quantity': True,
           'hypervisor_capacity': 0,
           'next_event_date': datetime.datetime(2024, 8, 3, 3, 59, 59, tzinfo=tzlocal()),
           'next_event_type': 'Subscription End',
           'product_name': 'Red Hat Enterprise Linux, Premium (One Year, '
                           'Enterprise Program)',
           'quantity': 1,
           'service_level': 'Premium',
           'sku': 'ESA0001',
           'subscriptions': [{'id': '14462076', 'number': '14462233'}],
           'total_capacity': 0,
           'usage': 'Production'}],
 'meta': {'count': 1, 'product': 'RHEL for x86', 'subscription_type': 'Annual'}}

In [24]: app.rhsm_subscriptions.rest_v2_client.subscriptions_api.get_sku_capacity_report(product_id)
2024-07-19 07:21:44.024 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=None, params=[]
Out[24]: 
{'data': [{'billing_provider': '',
           'has_infinite_quantity': True,
           'measurements': [0.0],
           'next_event_date': datetime.datetime(2024, 8, 3, 3, 59, 59, tzinfo=tzlocal()),
           'next_event_type': 'Subscription End',
           'product_name': 'Red Hat Enterprise Linux, Premium (One Year, '
                           'Enterprise Program)',
           'quantity': 1,
           'service_level': 'Premium',
           'sku': 'ESA0001',
           'subscriptions': [{'id': '14462076', 'number': '14462233'}],
           'usage': 'Production'}],
 'meta': {'count': 1,
          'measurements': ['Sockets'],
          'product': 'RHEL for x86',
          'subscription_type': 'Annual'}}
```
Openshift:
```
In [40]: app.rhsm_subscriptions.rest_client.subscriptions_api.get_sku_capacity_report("OpenShift Container Platform")
2024-07-19 09:29:24.238 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=None, params=[]
Out[40]: 
{'data': [{'billing_provider': '',
           'capacity': 0,
           'has_infinite_quantity': True,
           'hypervisor_capacity': 0,
           'next_event_date': datetime.datetime(2024, 8, 3, 3, 59, 59, tzinfo=tzlocal()),
           'next_event_type': 'Subscription End',
           'product_name': 'Red Hat OpenShift Container Platform, Premium (One '
                           'Year, Enterprise Program)',
           'quantity': 1,
           'service_level': 'Premium',
           'sku': 'ESA0012',
           'subscriptions': [{'id': '14462132', 'number': '14462289'}],
           'total_capacity': 0,
           'usage': ''},
          {'billing_provider': '',
           'capacity': 4,
           'has_infinite_quantity': False,
           'hypervisor_capacity': 0,
           'metric_id': 'Cores',
           'next_event_date': datetime.datetime(2024, 8, 3, 3, 59, 59, tzinfo=tzlocal()),
           'next_event_type': 'Subscription End',
           'product_name': 'Red Hat OpenShift Container Platform, Premium (2 '
                           'Cores or 4 vCPUs, Embedded, Billing)',
           'quantity': 2,
           'service_level': 'Premium',
           'sku': 'MCT3263',
           'subscriptions': [{'id': '14462175', 'number': '14462332'}],
           'total_capacity': 4,
           'uom': 'cores',
           'usage': ''}],
 'meta': {'count': 2,
          'product': 'OpenShift Container Platform',
          'subscription_type': 'Annual'}}

In [41]: 

In [41]: app.rhsm_subscriptions.rest_v2_client.subscriptions_api.get_sku_capacity_report("OpenShift Container Platform")
2024-07-19 09:29:32.834 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=None, params=[]
Out[41]: 
{'data': [{'billing_provider': '',
           'has_infinite_quantity': True,
           'measurements': [0.0, 0.0],
           'next_event_date': datetime.datetime(2024, 8, 3, 3, 59, 59, tzinfo=tzlocal()),
           'next_event_type': 'Subscription End',
           'product_name': 'Red Hat OpenShift Container Platform, Premium (One '
                           'Year, Enterprise Program)',
           'quantity': 1,
           'service_level': 'Premium',
           'sku': 'ESA0012',
           'subscriptions': [{'id': '14462132', 'number': '14462289'}],
           'usage': ''},
          {'billing_provider': '',
           'category': 'physical',
           'has_infinite_quantity': False,
           'measurements': [0.0, 4.0],
           'next_event_date': datetime.datetime(2024, 8, 3, 3, 59, 59, tzinfo=tzlocal()),
           'next_event_type': 'Subscription End',
           'product_name': 'Red Hat OpenShift Container Platform, Premium (2 '
                           'Cores or 4 vCPUs, Embedded, Billing)',
           'quantity': 2,
           'service_level': 'Premium',
           'sku': 'MCT3263',
           'subscriptions': [{'id': '14462175', 'number': '14462332'}],
           'usage': ''}],
 'meta': {'count': 2,
          'measurements': ['Sockets', 'Cores'],
          'product': 'OpenShift Container Platform',
          'subscription_type': 'Annual'}}
```

**Review by mmusil — APPROVED** [review]
LGTM when I tested the changes in EE and generated the API for iqe plugin. 
https://gitlab.cee.redhat.com/insights-qe/iqe-rhsm-subscriptions-plugin/-/merge_requests/796

But I'm not very familiar with the backend codebase.

**Comment by Sgitario** [comment]
Thanks @kahowell . The mapstruct changes look amazing to me +1. Merging.

**Pull request merged**
